### PR TITLE
fix(manager/poetry): Use buildpack installer

### LINKED ISSUE #19972: Setting a Poetry constraint causes lockfile update to fail
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Present on self-hosted and github app

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

Also observed on gitlab.com 15.8

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When setting a python and poetry constraint, the poetry lock update command fails because renovate attempts to pip install poetry using the wrong format for pip version constraints. See the repro here: https://github.com/WarpRat/renovate-poetry-contraint-repro/pull/6

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Failed to update poetry.lock file(branch="renovate/pylint-2.x")
{
  "err": {
    "name": "ExecError",
    "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\":\"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" docker.io/renovate/sidecar bash -l -c \"install-tool python 3.9.9 && pip install --user poetry1.3.2 && poetry update --lock --no-interaction pylint\"",
    "stderr": "ERROR: Could not find a version that satisfies the requirement poetry1.3.2 (from versions: none)\nERROR: No matching distribution found for poetry1.3.2\n",
    "stdout": "installing v2 tool python v3.9.9\nlinking tool python v3.9.9\nPython 3.9.9\npip 22.3.1 from /opt/buildpack/tools/python/3.9.9/lib/python3.9/site-packages/pip (python 3.9)\nInstalled v2 /usr/local/buildpack/tools/v2/python.sh in 24 seconds\nskip cleanup, not a docker build: 8d08c5ae1170\n",
    "options": {
      "cwd": "/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro",
      "encoding": "utf-8",
      "env": {
        "PIP_CACHE_DIR": "/tmp/renovate-cache/others/pip",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.local/bin:/home/ubuntu/bin:/opt/buildpack/tools/python/3.9.3/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/renovateapp/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "BUILDPACK_CACHE_DIR": "/tmp/containerbase",
        "CONTAINERBASE_CACHE_DIR": "/tmp/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\":\"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" docker.io/renovate/sidecar bash -l -c \"install-tool python 3.9.9 && pip install --user poetry1.3.2 && poetry update --lock --no-interaction pylint\"\nERROR: Could not find a version that satisfies the requirement poetry1.3.2 (from versions: none)\nERROR: No matching distribution found for poetry1.3.2\n",
    "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\":\"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/WarpRat/renovate-poetry-contraint-repro\" docker.io/renovate/sidecar bash -l -c \"install-tool python 3.9.9 && pip install --user poetry1.3.2 && poetry update --lock --no-interaction pylint\"\nERROR: Could not find a version that satisfies the requirement poetry1.3.2 (from versions: none)\nERROR: No matching distribution found for poetry1.3.2\n\n    at ChildProcess.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/exec/common.js:87:24)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)"
  }
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by WarpRat**
## Changes

A minor update to the format of the pip install of poetry when a version constraint is set.

## Context

- fixes #19972
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by WarpRat** [comment]
I'm struggling to confirm this is working in local testing but I'm not confident my testing is right either. I'll keep trying to confirm but if someone sees an obvious problem with this and wants to point me in the right direction I can update the MR

**Review by rarkins — CHANGES_REQUESTED** [review]
Although the existing code looks to be poorly checked as well, it's possible users are providing a constraint today, or perhaps even `===x.y.z`. The new code should check if it's a version or constraint and only append `===` if it's a version

**Comment by WarpRat** [comment]
> Although the existing code looks to be poorly checked as well, it's possible users are providing a constraint today, or perhaps even `===x.y.z`. The new code should check if it's a version or constraint and only append `===` if it's a version

I didn't even consider that 🤦 - defining our constraint that way probably solves our issue at the moment but I'll work on getting this pull request updated to do better checking for other folks who make the same mistake as we did. Thanks for the review.

File: lib/modules/manager/poetry/artifacts.ts [hunk 1]
```diff
@@ -187,7 +187,9 @@ export async function updateArtifacts({
       docker: {},
       toolConstraints: [{ toolName: 'python', constraint }],
       preCommands: [
-        `pip install --user ${quote(`poetry${poetryVersion ?? ''}`)}`,
+        `pip install --user ${quote(
+          `poetry${poetryVersion ? `===${poetryVersion}` : ''}`
+        )}`,
```
Comment (viceice) [inline lib/modules/manager/poetry/artifacts.ts:192]: convert to tool constraints instead, it's now supported by underlying containerbase image

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/poetry/artifacts.ts (hunk 1 above)
Comment (WarpRat) [reply lib/modules/manager/poetry/artifacts.ts]: Thank you for the tip, that's much cleaner. I've tested locally with `binarySource=install` and it works as expected.

**Review by viceice — CHANGES_REQUESTED** [review]
tests need an update

**Comment by WarpRat** [comment]
> tests need an update

Ah, sorry, should have run those locally before requesting a re-review. I'll find some time to make those updates this week. Thanks again.

**Comment by WarpRat** [comment]
Thanks for taking care of this @rarkins - I got busy at work and never had a chance to come back and look at the tests. Sorry to leave it dangling.

**Pull request closed without merging**